File: src/app/not-found.ts
import { DEFAULT_REPLY_MARKUP } from "../constant";
import type { TelegramRequest, TelegramResponse } from "../types";

export function notFound(req: TelegramRequest): TelegramResponse {
  if (req.message) {
    return {
      method: "sendMessage",
      chat_id: req.message.chat.id,
      parse_mode: "HTML",
      text: "<i>Unknown command</i>",
      reply_markup: DEFAULT_REPLY_MARKUP,
    };
  } else if (req.callback_query) {
    return {
      method: "editMessageText",
      message_id: req.callback_query.message!.message_id,
      chat_id: req.callback_query.from.id,
      parse_mode: "HTML",
      text: "<i>Invalid session</i>",
    };
  }

  throw new Error("Unknown request type");
}

File: internal/handler/not_found.go
package handler

import (
	"context"

	"github.com/fidrasofyan/version-watcher-bot/internal/custom_error"
	"github.com/fidrasofyan/version-watcher-bot/internal/repository"
	"github.com/fidrasofyan/version-watcher-bot/internal/service"
	"github.com/fidrasofyan/version-watcher-bot/internal/types"
)

func NotFound(ctx context.Context, req types.TelegramUpdate) (*types.TelegramResponse, error) {
	// Is it callback query?
	if req.CallbackQuery.Id != "" {
		// Delete chat
		err := repository.TelegramDeleteChat(ctx, req.CallbackQuery.From.Id)
		if err != nil {
			return nil, custom_error.NewError(err)
		}

		// Answer callback query
		err = service.AnswerCallbackQuery(ctx, &service.AnswerCallbackQueryParams{
			CallbackQueryId: req.CallbackQuery.Id,
		})
		if err != nil {
			return nil, custom_error.NewError(err)
		}

		return &types.TelegramResponse{
			Method:    types.TelegramMethodEditMessageText,
			MessageId: req.CallbackQuery.Message.MessageId,
			ChatId:    req.CallbackQuery.Message.Chat.Id,
			ParseMode: types.TelegramParseModeHTML,
			Text:      "<i>Invalid session</i>",
		}, nil
	}

	return &types.TelegramResponse{
		Method:      types.TelegramMethodSendMessage,
		ChatId:      req.Message.Chat.Id,
		ParseMode:   types.TelegramParseModeHTML,
		Text:        "<i>Unknown command</i>",
		ReplyMarkup: types.DefaultReplyMarkup,
	}, nil
}
